linker-utils/src/aarch64.rs

```diff
@@ -100,522 +101,728 @@ impl RelaxationKind {
 
 #[must_use]
 pub const fn relocation_type_from_raw(r_type: u32) -> Option<RelocationKindInfo> {
-    let (kind, size, mask) = match r_type {
+    let (kind, size, mask, range) = match r_type {
         // 5.7.4   Static miscellaneous relocations
-        object::elf::R_AARCH64_NONE => (RelocationKind::None, RelocationSize::ByteSize(0), None),
+        object::elf::R_AARCH64_NONE => (
+            RelocationKind::None,
+            RelocationSize::ByteSize(0),
+            None,
+            AllowedRange::no_check(),
+        ),
 
         // 5.7.5   Static Data relocations
         // Data relocations
-        object::elf::R_AARCH64_ABS64 => {
-            (RelocationKind::Absolute, RelocationSize::ByteSize(8), None)
-        }
-        object::elf::R_AARCH64_ABS32 => {
-            (RelocationKind::Absolute, RelocationSize::ByteSize(4), None)
-        }
-        object::elf::R_AARCH64_ABS16 => {
-            (RelocationKind::Absolute, RelocationSize::ByteSize(2), None)
-        }
-        object::elf::R_AARCH64_PREL64 => {
-            (RelocationKind::Relative, RelocationSize::ByteSize(8), None)
-        }
-        object::elf::R_AARCH64_PREL32 => {
-            (RelocationKind::Relative, RelocationSize::ByteSize(4), None)
-        }
-        object::elf::R_AARCH64_PREL16 => {
-            (RelocationKind::Relative, RelocationSize::ByteSize(2), None)
-        }
+        object::elf::R_AARCH64_ABS64 => (
+            RelocationKind::Absolute,
+            RelocationSize::ByteSize(8),
+            None,
+            AllowedRange::no_check(),
+        ),
+        object::elf::R_AARCH64_ABS32 => (
+            RelocationKind::Absolute,
+            RelocationSize::ByteSize(4),
+            None,
+            AllowedRange::new(-(2i64.pow(31)), 2i64.pow(32)),
+        ),
+        object::elf::R_AARCH64_ABS16 => (
+            RelocationKind::Absolute,
+            RelocationSize::ByteSize(2),
+            None,
+            AllowedRange::new(-(2i64.pow(15)), 2i64.pow(16)),
```

Comment:
Why are many of these ranges so asymmetrical? e.g. this one is [-32768..65536)

Reply:
Dunno, I just followed the SPEC:
https://github.com/ARM-software/abi-aa/blob/main/aaelf64/aaelf64.rst

Should we file an issue against the documentation? It might be the case, they wrongly placed `-(2^15)` instead of `-(2^16 - 1)`.

Reply:
Ah, I see. The spec says:

> These overflow ranges permit either signed or unsigned narrow values to be created from the intermediate result viewed as a 64-bit signed integer. If the place is intended to hold a narrow signed value and INTn_MAX < X <= UINTn_MAX, no overflow will be detected but the positive result will be interpreted as a negative value.

So we could check a tighter bound if we knew that the value was signed, but we don't.